market_maker_keeper/leverj_market_maker_keeper.py

```diff
@@ -0,0 +1,244 @@
+
+        parser.add_argument("--refresh-frequency", type=int, default=3,
+                            help="Order book refresh frequency (in seconds, default: 3)")
+
+        parser.add_argument("--pair", type=str, required=True,
+                            help="Token pair (sell/buy) on which the keeper will operate")
+
+        parser.add_argument("--debug", dest='debug', action='store_true',
+                            help="Enable debug output")
+
+        self.arguments = parser.parse_args(args)
+    
+        if "infura" in self.arguments.rpc_host:
+            self.web3 = Web3(HTTPProvider(endpoint_uri=f"http://{self.arguments.rpc_host}",
+                                      request_kwargs={"timeout": self.arguments.rpc_timeout}))
+        else:
+            self.web3 = Web3(HTTPProvider(endpoint_uri=f"http://{self.arguments.rpc_host}:{self.arguments.rpc_port}",
+                                      request_kwargs={"timeout": self.arguments.rpc_timeout}))
+
+        self.web3.eth.defaultAccount = self.arguments.eth_from
+        register_keys(self.web3, self.arguments.eth_key)
+
+        setup_logging(self.arguments)
+
+        self.bands_config = ReloadableConfig(self.arguments.config)
+        self.price_feed = PriceFeedFactory().create_price_feed(self.arguments)
+        self.spread_feed = create_spread_feed(self.arguments)
+        self.control_feed = create_control_feed(self.arguments)
+        self.order_history_reporter = create_order_history_reporter(self.arguments)
+
+        self.history = History()
+
+        self.leverj_api = LeverjAPI(web3=self.web3,
+                                    api_server=self.arguments.leverj_api_server,
+                                    account_id=self.arguments.account_id,
+                                    api_key=self.arguments.api_key,
+                                    api_secret=self.arguments.api_secret,
+                                    timeout=self.arguments.leverj_timeout)
+
+
+        self.order_book_manager = OrderBookManager(refresh_frequency=self.arguments.refresh_frequency)
+        self.order_book_manager.get_orders_with(lambda: self.leverj_api.get_orders(self.pair()))
+        self.order_book_manager.get_balances_with(lambda: self.leverj_api.get_balances())
+        self.order_book_manager.cancel_orders_with(lambda order: self.leverj_api.cancel_order(order.order_id))
+        self.order_book_manager.enable_history_reporting(self.order_history_reporter, self.our_buy_orders,
+                                                         self.our_sell_orders)
+        self.order_book_manager.start()
+
+    def main(self):
+        with Lifecycle() as lifecycle:
+            lifecycle.initial_delay(1)
+            lifecycle.on_startup(self.startup)
+            lifecycle.every(1, self.synchronize_orders)
+            lifecycle.on_shutdown(self.shutdown)
+
+    def startup(self):
```

Comment:
can you explain the difference between the precision here and the precision you get / use in pyexchange?

Reply:
so we have extra precision in market maker keeper to do some calculations, but finally, when we send the order in pyexchange we limit it to the precision of the exchange.